### PR TITLE
ZIO Test: Further Improvements to MockClock Handling of Sleep

### PR DESCRIPTION
**Pull request by adamgfraser**
As suggested by @jozic in #1447, adds an additional example involving resetting the `MockClock` to test an effect recurring on a fixed schedule. Also adds examples to test suite to verify that they work correctly and prevent regressions.

### CODE REVIEW

**Comment by ghostdogpr** [comment]
Seems like the new test fails in most builds? Actually there’s no detail on which test fails :/

**Comment by adamgfraser** [comment]
Yes I am reworking this. Unfortunately the test suite for ZIO Test doesn’t have nice error messages. May have to clean that up at some point.

**Comment by adamgfraser** [comment]
After spending more time on this, I came to the conclusion that the previous `MockClock` implementation, while an improvement, was not sufficient.

While it was effective for testing some time related applications such as `timeout`, the fact that `sleep` always scheduled effects for an absolute time from zero time instead of relative time from some notion of the "current time" made it inconsistent with the actual `Clock` and created issues with testing code involving repetitive effects. For example, under the old implementation if we had an effect that executed repeatedly every hour and advanced the `MockClock` by an hour, further repetitions of that effect would execute without any further delay, requiring further coordination mechanisms such as latches that did not provide a good user experience.

This PR further improves `MockClock` so that `sleep` schedules actions for a relative time in the future. We avoid race conditions between calling `sleep` and updating the time by using a notion of "fiber time", backed by a `FiberRef`, that is incremented for each fiber when it sleeps and used as a relative basis for scheduling `sleep` calls, and "clock time" controlled by the user. With this we can test sophisticated concepts involving time in a straightforward way. For example, here is how we could test an effect that performs an action at a fixed interval:

```scala
val io = for {
  q <- Queue.unbounded[Unit]
  _ <- (q.offer(()).delay(60.minutes)).forever.fork
  a <- q.poll.map(_.isEmpty)
  _ <- MockClock.adjust(60.minutes)
  b <- q.take.as(true)
  c <- q.poll.map(_.isEmpty)
  _ <- MockClock.adjust(60.minutes)
  d <- q.take.as(true)
  e <- q.poll.map(_.isEmpty)
} yield a && b && c && d && e
```

@jdegoes Could I get a review on this when you have the chance?

**Comment by iravid** [comment]
Very clever use of `FiberRef` @adamgfraser!

File: test/shared/src/test/scala/zio/test/TestUtils.scala [hunk 1]
```diff
@@ -12,6 +14,18 @@ object TestUtils {
         (p, fail(s))
     }.recover { case _ => (false, fail(s)) }
 
+  def nonFlaky[R, E](test: ZIO[R, E, Boolean]): ZIO[R, E, Boolean] = {
```
Comment (jdegoes) [inline test/shared/src/test/scala/zio/test/TestUtils.scala]: Implementation could be: `test.repeat(Schedule.recurs(100) *> Schedule.identity[Boolean])`

Comment (adamgfraser) [reply test/shared/src/test/scala/zio/test/TestUtils.scala]: Thanks! Much nicer! Now if we can only get rid of that pesky `Clock` type constraint.

Comment (jdegoes) [reply test/shared/src/test/scala/zio/test/TestUtils.scala]: Which we can shortly, I hope, with #1348!

**Pull request merged**